src/Control/Carrier/Telemetry.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Control.Carrier.Telemetry (
+  TelemetryC,
+  IgnoreTelemetryC,
+  withTelemetry,
+  withoutTelemetry,
+  bracket',
+) where
+
+import Control.Concurrent.STM (atomically, putTMVar, tryReadTMVar)
+import Control.Concurrent.STM.TBMQueue (tryWriteTBMQueue)
+import Control.Effect.Exception (Exception (fromException), catch, mask, throwIO)
+import Control.Effect.Lift (Lift, sendIO)
+import Control.Effect.Telemetry (Telemetry (..))
+import Control.Monad (void, when)
+import Control.Monad.IO.Class (MonadIO (..))
+import Data.Maybe (isNothing)
+
+import App.Fossa.Telemetry.Types (
+  TelemetryCtx (..),
+  TelemetryTimeSpent (TelemetryTimeSpent),
+  TimedLogRecord (TimedLogRecord),
+ )
+import App.Fossa.Telemetry.Utils (
+  mkTelemetryCtx,
+  mkTelemetryRecord,
+ )
+import Control.Exception qualified as Exc
+import Data.Time.Clock (getCurrentTime)
+import Data.Tracing.Instrument (incCount)
+import System.Exit (ExitCode (ExitSuccess))
+
+import App.Fossa.Telemetry.Sink.Common (emitTelemetry)
+import Control.Algebra (Algebra (alg), Has, type (:+:) (L, R))
+import Control.Carrier.Reader (ReaderC, ask, runReader)
+import System.TimeIt (timeItT)
+
+newtype TelemetryC m a = TelemetryC {runTelemetryC :: ReaderC TelemetryCtx m a}
+  deriving (Applicative, Functor, Monad, MonadFail, MonadIO)
+
+instance (Algebra sig m, MonadIO m, Has (Lift IO) sig m) => Algebra (Telemetry :+: sig) (TelemetryC m) where
+  alg hdl sig ctx = case sig of
+    L op -> do
+      telCtx <- TelemetryC (ask @(TelemetryCtx))
+      case op of
+        SetTelemetrySink sink -> do
+          maybeSink <- sendIO $ atomically $ tryReadTMVar (telSink telCtx)
+          when (isNothing maybeSink) $
+            sendIO $ atomically $ putTMVar (telSink telCtx) sink
+          ctx <$ pure ()
+        TrackUsage feat -> do
+          ctx <$ (sendIO . atomically $ incCount feat $ telCounters telCtx)
+        TrackConfig cmd cfg -> do
+          maybeCfg <- sendIO $ atomically $ tryReadTMVar (telFossaConfig telCtx)
+          case maybeCfg of
+            Nothing -> ctx <$ sendIO (atomically $ putTMVar (telFossaConfig telCtx) (cmd, cfg))
+            Just _ -> ctx <$ pure ()
+        TrackTimeSpent computeName act -> do
+          (timeTook, res) <- timeItT (hdl (act <$ ctx))
```

Comment:
- [ ] Should we be capturing real time as opposed to cpu time?

Reply:
- [ ] We can do this better, we should be doing `trackT :: TelemetryC m a -> TelemetryC m a` in taking complexity out of algebra

Reply:
I'm not against capturing CPU time, but if we can capture real-time I think that would be more valuable

Reply:
We should prefer wall-clock time to CPU time, if we have to choose.  Wall clock time tells us about the user's experience, CPU time tells us the machine's experience.

Reply:
Ideally, we should have both.

Reply:
I think we should have both CPU time and wall clock time if that is an option. I was debugging an issue with Rodrigo yesterday related to the CLI and without both wall clock time and CPU time we would have been stuck.

Reply:
I have changed it to observed time via system clocks in https://github.com/fossas/fossa-cli/pull/831/commits/a61faf4041aefec21439c8d37756929a5eb386f0, but we can add CPU time in subsequent PR (I want to keep this PR from exploding than it already is)